File: freshman year/CSE12/hw7/Driver.java
/****************************************************************************

                                                        Jamie Zhang
                                                        CSE12, WI15
                                                        Feb 15, 2015
                                                        cs12xre
                                Assignment Seven

File Name:      Driver.java
Description:    This file contains the driver as well as code for a 
                UCSDStudent object. The driver is used to run and test the 
		code of the tree and the UCSDStudent is the test object that
		will go into the tree.
****************************************************************************/

import java.io.*;

/*************************************************************************** 
 * Class Name: UCSDStudent
 * Description: A UCSDStudent object used to test the tree. Has name
 *              and student number fields.
 * Data Fields: String name - the name of the student
 *              long studentnum - the student number of the student
 * Public Methods: getName - returns the name of the student
 *                 equals - checks if this UCSDStudent is equal to 
 *                          another object
 *                 isGreaterThan - checks if this UCSDStudent's name is 
 *                                 greater in the alphabet to another name
 *                 toString - returns a string representation of the object
 ***************************************************************************/
class UCSDStudent extends Base {

   private String name; //the name of the student
   private long studentnum; //the student number


   /**
   * Constructor. 
   * Allocates and initializes the elements of the UCSDStudent
   *
   * @param    nm    the name of the UCSDStudent
   * @param    num   the student number of the UCSDStudent
   */   
   public UCSDStudent(String nm, long num)
   {
      this.name = new String(nm);
      studentnum = num;
   }


   /**
   * Returns the name.
   *
   * @return name of the UCSDStudent
   */   
   public String getName()
   {
      return this.name;
   }

 
   /**
   * Checks for equality. Two items are equal if their names are the same.
   *
   * @return boolean - T/F depending on if the names are equal or not
   */   
   public boolean equals(Object o)
   {	   
      if(o == null || !(o instanceof UCSDStudent))
      {
         return false; //not equal if o is null or o isn't a UCSDStudent		      
      }
      else
      {
         if(this.getName().equals(((UCSDStudent)o).getName()))
	 {
            return true; //true if the names match
	 }
	 else
	 {
            return false;
	 }
      }
   }


   /**
   * Checks the relation of two names.
   *
   * @return boolean - T/F depending on the relation between the two names                 
   */      
   public boolean isGreaterThan(Base base)
   {
      if(name.compareTo(base.getName()) > 0)
      {
	 //true if this UCSDStudent's name is greater in the alphabet than 
	 //the other name     
         return true;
      }
      else
      {
         //false if this UCSDStudent's name is less in the alphabet than 
	 //the other name	      
         return false;
      }
   }


   /**
   * Creates a string representation of the UCSDStudent
   *
   * @return  String representation of the UCSDStudent
   */   
   public String toString () {
      return "name:  " + name + "  studentnum:  " + studentnum;
   }
}

public class Driver {
   private static final short NULL = 0;

   public static void main (String [] args) {
        
      /* initialize debug states */
      Tree.debugOff();

      /* check command line options */
      for (int index = 0; index < args.length; ++index) {
         if (args[index].equals("-x"))
            Tree.debugOn();
      }


      /* The real start of the code */
      SymTab<UCSDStudent> symtab = 
                          new SymTab<UCSDStudent>("UCSDStudentTree");
      String buffer = null;
      char command;
      long number = 0;

      System.out.println ("Initial Symbol Table:\n" + symtab);

      while (true) {
         command = NULL; // reset command each time in loop
         System.out.print ("Please enter a command:  " + 
                           "((a)llocate, (i)nsert, (l)ookup, " +
                           "(r)emove, (w)rite):  ");

         try {
            command = MyLib.getchar ();
            MyLib.clrbuf (command); // get rid of return

            switch (command) {
               case 'a':
                  System.out.print
                              ("Please enter name of new Tree to " +
                               "allocate:  ");
                                
                  buffer = MyLib.getline ();// formatted input
                  symtab = new SymTab<UCSDStudent>(buffer);
                  break;

               case 'i':
                  System.out.print
                         ("Please enter UCSD student name to insert:  ");

                  buffer = MyLib.getline ();// formatted input

                  System.out.print
                          ("Please enter UCSD student number:  ");

                  number = MyLib.decin ();
                  MyLib.clrbuf (command); // get rid of return

                  // create student and place in symbol table
                  symtab.insert(new UCSDStudent (buffer, number));
                  break;

               case 'l': { 
		  UCSDStudent found;      // whether found or not

                  System.out.print
                       ("Please enter UCSD student name to lookup:  ");
                  buffer = MyLib.getline ();// formatted input

                  UCSDStudent stu = new UCSDStudent (buffer, 0);
                  found = symtab.lookup (stu);
                                
                  if (found != null) {
                     System.out.println("Student found!");
                     System.out.println(found);
                  }
                  else
                     System.out.println ("student " + buffer
                                         + " not there!");
                  }
                  break;
                        
               case 'r': { 
                  UCSDStudent removed; // data to be removed

                  System.out.print
                     ("Please enter UCSD student name to remove:  ");

                  buffer = MyLib.getline ();

                  UCSDStudent stu = new UCSDStudent (buffer, 0);

                  removed = symtab.remove(stu);

                  if (removed != null) {
                     System.out.println("Student removed!"); 
                     System.out.println(removed);
                  }
                  else
                     System.out.println ("student " + buffer
                                         + " not there!");
                  }
                  break;

               case 'w':
                  System.out.println("The Symbol Table " +
                                     "contains:\n" + symtab);
            }
         }
         catch (EOFException eof) {
            break;
         }
      }
      System.out.println("\nFinal Symbol Table:\n" + symtab);
   }
}

File: freshman year/CSE12/hw7/Driver.c
/****************************************************************************

                                                        Jamie Zhang
                                                        CSE12, WI15
                                                        Feb 15, 2015
                                                        cs12xre
                                Assignment Seven

File Name:      Driver.c
Description:    This file contains the driver as well as code for a 
                UCSDStudent object. The driver is used to run and test the 
		code of the tree and the UCSDStudent is the test object that
		will go into the tree.
****************************************************************************/

#include <cstdlib>
#include <iostream>
#include <string.h>
#include "Base.h"
#include "SymTab.h"
using namespace std;

#ifdef NULL
#undef NULL
#define NULL 0
#endif

class UCSDStudent : public Base {
   char * name; //name of the student
   long studentnum; //student number of the student
public:
   /*-----------------------------------------------------------------------
   Name:                  UCSDStudent
   Purpose:               Constructor for the UCSDStudent.
   Description:           Allocates and initializes the UCSDStudent's data 
                          fields.
   Input:                 nm - name of the student
                          val - student number of the student
   -----------------------------------------------------------------------*/
   UCSDStudent(char * nm, long val = 0):
      name(strdup(nm)), studentnum(val) {}


   /*-----------------------------------------------------------------------
   Name:                  UCSDStudent
   Purpose:               Copy constructor for the UCSDStudent.
   Description:           Allocates and initializes the UCSDStudent's data 
                          fields based on an already existing UCSDStudent.
			  Makes a deep copy.
   Input:                 stu - an already existing UCSDStudent to copy
   -----------------------------------------------------------------------*/   
   UCSDStudent(const UCSDStudent & stu)
   {
      //initializes data fields by making a deep copy of stu's data	   
      name = strdup(stu.name);
      studentnum = stu.studentnum;
   }


   /*-----------------------------------------------------------------------
   Name:                  ~UCSDStudent
   Purpose:               Destructor for the UCSDStudent.
   Description:           Frees up the memory allocated for the UCSDStudent's 
                          name.
   Input:                 none
   -----------------------------------------------------------------------*/    
   ~UCSDStudent(void)
   {
      free(name); //deallocates memory
   }


   /*-----------------------------------------------------------------------
   Function Name:         operator char *
   Purpose:               Gets the name of the UCSDStudent.
   Description:           Returns the name of the UCSDStudent.
   Input:                 void - no input
   Result:                char * - the name of the UCSDStudent
   -----------------------------------------------------------------------*/   
   operator char * (void) const
   {
      return name;
   }


   /*-----------------------------------------------------------------------
   Function Name:         operator ==
   Purpose:               Checks to see if two UCSDStudent objects are equal.
   Description:           Uses strcmp to see if the names of two UCSDStudents
                          are the same.
   Input:                 bbb - what we're comparing with
   Result:                long - 1 if equal, 0 if not equal
   -----------------------------------------------------------------------*/     
   long operator == (const Base & bbb) const
   {
      return !strcmp(name, bbb);
   }


   /*-----------------------------------------------------------------------
   Function Name:         operator >
   Purpose:               Checks if the name of this UCSDStudent is greater in
                          the alphabet than the name of another UCSDStudent.
   Description:           Uses strcmp to check the relation between two
                          UCSDStudent's names.
   Input:                 bbb - what we're comparing with
   Result:                long - 1 if greater than, 0 if less than
   -----------------------------------------------------------------------*/     
   long operator > (const Base & bbb) const
   {
      return (strcmp(name, bbb) > 0) ? 1 : 0;
   }


   ostream & Write (ostream & stream) const {
      return stream << "name:  " << name << "  studentnum:  " << studentnum; 
   }
};

int main (int argc, char * const * argv) {
   char buffer[80];
   char command;
   long number;

   Tree::Set_Debug(0);

   if (argc != 1 && strcmp("-x", argv[1]) == 0) {
      Tree::Set_Debug(1);
   }
	
   SymTab * ST;
   ST = new SymTab ("UCSDStudentTree");
   ST->Write (cout << "Initial Symbol Table:\n");

   while (cin) {
      command = NULL;		// reset command each time in loop
      cout <<  "Please enter a command:  ((a)llocate, (i)nsert, " <<
			"(l)ookup, (r)emove, (w)rite):  ";
      cin >> command;

      switch (command) {

         case 'a':
	    cout << "Please enter name of new Tree to allocate:  ";
	    cin >> buffer; 

	    delete ST;
	    ST = new SymTab (buffer);

	    break;

         case 'i':
	    cout << "Please enter UCSD student name to insert:  ";
	    cin >> buffer;	// formatted input

	    cout << "Please enter UCSD student number:  ";
	    cin >> number;

	    // create student and place in symbol table
	    ST->Insert (new UCSDStudent (buffer, number));
	    break;

	 case 'l': {
	    const Base * found;	// whether found or not

	    cout << "Please enter UCSD student name to lookup:  ";
	    cin >> buffer;	// formatted input

	    UCSDStudent stu (buffer, 0);
	    found = ST->Lookup (&stu);
			
	    if (found)
	       found->Write (cout << "Student found!\n") << "\n";
	    else
	       cout << "student " << buffer << " not there!\n";
	    }
	    break;
		
	 case 'r': {
	    Base * removed;	// data to be removed

	    cout << "Please enter UCSD student name to remove:  ";
	    cin >> buffer;	// formatted input

	    UCSDStudent stu (buffer, 0);
	    removed = ST->Remove (&stu);

	    if (removed)
	       removed->Write (cout << "Student removed!\n") << "\n";
	    else
	       cout << "student " << buffer << " not there!\n";
	    }
	    break;


	 case 'w':
	    ST->Write (cout << "The Symbol Table contains:\n");
      }
   }

   ST->Write (cout << "\nFinal Symbol Table:\n");
   delete ST;
}
